sycl/source/detail/program_impl.cpp

```diff
@@ -38,14 +39,34 @@ program_impl::program_impl(
     throw runtime_error("Non-empty vector of programs expected",
                         PI_INVALID_VALUE);
   }
+
+  // Sort the programs to avoid deadlocks due to locking multiple mutexes &
+  // verify that all programs are unique.
+  std::sort(ProgramList.begin(), ProgramList.end(),
+            [](const shared_ptr_class<program_impl> &A,
+               const shared_ptr_class<program_impl> &B) {
+              return A.get() < B.get();
+            });
+  auto It = std::unique(ProgramList.begin(), ProgramList.end(),
```

Comment:
I believe lambda is not needed here 'cause `shared_ptr` has [comparison operators](https://en.cppreference.com/w/cpp/memory/shared_ptr/operator_cmp)